pkg/bpf/map.go

```diff
@@ -884,3 +890,61 @@ func (m *Map) CheckAndUpgrade(desired *MapInfo) bool {
 		desired.Flags,
 	)
 }
+
+// mapTypeToFeatureString maps a MapType into a string defined by run_probes.sh
+func mapTypeToFeatureString(mt MapType) string {
+	var featureString string
+	switch mt {
+	case MapTypeLPMTrie:
+		featureString = fmt.Sprintf("#define HAVE_LPM_MAP_TYPE")
+	case MapTypeLRUHash:
+		featureString = fmt.Sprintf("#define HAVE_LRU_MAP_TYPE")
+	default:
+		break
+	}
+	return featureString
+}
+
+// ReadFeatureProbes reads the bpf_features.h file at the specified path (as
+// generated by bpf/run_probes.sh), and stores the results of the kernel
+// feature probing.
+func ReadFeatureProbes(filename string) {
+	f, err := os.Open(filename)
+	if err != nil {
+		// Should not happen; the caller ensured that the file exists
+		log.Fatal(err)
+	}
+	defer f.Close()
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		for mapType := MapTypeHash; mapType < MapTypeMaximum; mapType++ {
+			featureString := mapTypeToFeatureString(mapType)
+			if featureString != "" &&
+				bytes.Compare(scanner.Bytes(), []byte(featureString)) == 0 {
+				log.Debugf("Detected support for map type %s", mapType.String())
+				supportedMapTypes[mapType] = true
+			}
+		}
+	}
+
+	for mapType := MapTypeHash; mapType < MapTypeMaximum; mapType++ {
+		if mapTypeToFeatureString(mapType) == "" {
+			log.Debugf("Skipping support detection for map type %s", mapType.String())
+		} else if _, probed := supportedMapTypes[mapType]; !probed {
+			log.Debugf("Detected no support for map type %s", mapType.String())
+			supportedMapTypes[mapType] = false
+		}
+	}
+}
+
+// SupportedMapType converts the specified MapType into a supported MapType for
+// the current kernel.
+//
+// If the kernel supports the specified MapType, it returns the MapType.
+// Otherwise, it determines the fallback MapType and returns that type.
+func SupportedMapType(mapType MapType) MapType {
```

Comment:
I can't think of a good name, would you be happy with a Fatalf() instead since it should be developers who hit this kind of case?

```
@@ -943,8 +945,12 @@ func ReadFeatureProbes(filename string) {
 // If the kernel supports the specified MapType, it returns the MapType.
 // Otherwise, it determines the fallback MapType and returns that type.
 func SupportedMapType(mapType MapType) MapType {
-       if supported, probed := supportedMapTypes[mapType]; supported || !probed {
+       supported, probed := supportedMapTypes[mapType]
+       if supported {
                return mapType
+       } else if probed {
+               return MapTypeHash
        }
-       return MapTypeHash
+
+       log.Fatalf("Map type %s not supported by feature probes.", mapType.String())
 }
```

Reply:
I would call it `GetLRUMapType()` which returns either LRU or Hash and the comment clearly stating that the function is meant to be used when the datapath uses LRU on best effort basis.